### PR TITLE
Add support for 2D and 3D Wire fillet

### PR DESCRIPTION
**Pull request by dov**
Add support for 2D and 3D Wire fillet through OCCT ChFi2d_FilletAPI

### CODE REVIEW

**Comment by adam-urbanczyk** [comment]
Hey, thanks, but could you _not_ force push

**Comment by dov** [comment]
@adam-urbanczyk 
Regarding force push. Sorry, I was (ab-)using it in order to get the CI support. I'll make sure in the future to clean all the CI errors before creating pull requests.

Regarding the PR itself, I added a vertices parameter, but in terms of a list of vertex indices. I was wondering whether it might be better to give the coordinates. What do you think?

**Comment by adam-urbanczyk** [comment]
Every push will trigger CI, am I missing something?

File: cadquery/occ_impl/shapes.py [hunk 1]
```diff
@@ -2379,6 +2381,65 @@ def chamfer2D(self, d: float, vertices: Iterable[Vertex]) -> "Wire":
 
         return f.chamfer2D(d, vertices).outerWire()
 
+    def fillet(self, radius: float, vertices: Optional[Iterable[int]] = None) -> "Wire":
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/shapes.py:2384]: I'd propose to use `Optional[Iterable[Vertex]]` here

File: cadquery/occ_impl/shapes.py [hunk 2]
```diff
@@ -2379,6 +2381,75 @@ def chamfer2D(self, d: float, vertices: Iterable[Vertex]) -> "Wire":
...
+
+        :return: A wire with filleted corners
+        """
+        edges = list(self)
+        newEdges = []
+        currentEdge = edges[0]
+
+        # Make vertices into a set so that we can query for membership
+        # in O(log n).
+        if vertices is not None:
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/shapes.py:2403]: you should simply use vertices as-is an convert them to set

File: cadquery/occ_impl/shapes.py [hunk 3]
```diff
@@ -2379,6 +2381,75 @@ def chamfer2D(self, d: float, vertices: Iterable[Vertex]) -> "Wire":
...
+
+            # Create a plane that is spanned by currentEdge and nextEdge
+            currentDir = currentEdge.tangentAt(1)
+            nextDir = nextEdge.tangentAt(0)
+            normalDir = currentDir.cross(nextDir)
+
+            # Check conditions for skipping fillet:
+            #  1. The edges are are parallell
+            #  2. The vertex is not in the vertices white list
+            vertex = currentEdge.startPoint().toTuple()
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/shapes.py:2417]: remove `toTuple()`

File: cadquery/occ_impl/shapes.py (hunk 2 above)
Comment (dov) [reply cadquery/occ_impl/shapes.py:2403]: Sorry. I don't understand.  `Vertex` isn't immutable, and you can't put it into python set. Which is why I'm doing the casting to an immutable tuple. Did I miss something?

Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py]: I applied my suggestions, let's see if it works

Comment (dov) [reply cadquery/occ_impl/shapes.py]: I checked the immutability aspect of Vertex and it seems I was wrong. It's just a wrapper for `TopoDS_Vertex` which seems to be immutable. So using sets on Vertex is ok. 

However using vertices directly without casting to Vertex or tuple will fail if the user passes a list for coordinates to vertices.

I.e. the following code worked before your change, but will fail after your change:

```
        points = [
            [0.000, 0.000, 0.000],
            [-0.287, 1.183, -0.592],
            [-1.404, 4.113, -2.787],
            [-1.332, 1.522, 0.553],
            [7.062, 0.433, -0.097],
            [8.539, -0.000, -0.000],
        ]
        wire = Wire.makePolygon(points, close=False)

        # Fillet a single vertex
        wfillet = wire.fillet(radius=0.560, vertices=[points[1]])
        assert len(wfillet.Edges()) == len(points)

```

Comment (dov) [reply cadquery/occ_impl/shapes.py]: I'd rather do a potentially redundant cast in the implementation to catch a casual misuse of the API than having the code raise a cryptic exception. But if it's fine with you, I'll yield.

Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py]: The API is specified by the type annotations, IMO nothing is cryptic here. The general vision is that you use selectors to specify fillets and chamfers.

**Review by adam-urbanczyk — APPROVED** [review]

**Comment by dov** [comment]
:partying_face:

**Comment by adam-urbanczyk** [comment]
FYI: in the end I refactored `Wire.Vertices()` so that the rv is ordered.

File: cadquery/occ_impl/shapes.py [hunk 4]
```diff
@@ -2379,6 +2381,90 @@ def chamfer2D(self, d: float, vertices: Iterable[Vertex]) -> "Wire":
...
+        for i in range(len(edges) - 1):
+            nextEdge = edges[i + 1]
+
+            # Create a plane that is spanned by currentEdge and nextEdge
+            currentDir = currentEdge.tangentAt(1)
+            nextDir = nextEdge.tangentAt(0)
+            normalDir = currentDir.cross(nextDir)
+
+            # Check conditions for skipping fillet:
+            #  1. The edges are are parallell
```
Comment (jmwright) [inline cadquery/occ_impl/shapes.py:2411]:
```suggestion
            #  1. The edges are are parallel
```

**Review by jmwright — APPROVED** [review]
Thanks @dov there is just the one comment suggestion from me.

**Comment by dov** [comment]
> Thanks @dov there is just the one comment suggestion from me.

Thanks @jmwright . I fixed it and also removed another typo in the comment.

**Comment by adam-urbanczyk** [comment]
Thanks @dov !

**Pull request merged**